File: lab3/lab3-1.py
class Solution:
    def majorityElement(self, nums):
        def majority_element_rec(left, right):
            if left == right:
                return nums[left]
            
            mid = (left + right) // 2
            left_majority = majority_element_rec(left, mid)
            right_majority = majority_element_rec(mid + 1, right)
            
            if left_majority == right_majority:
                return left_majority
            
            left_count = sum(1 for i in range(left, right + 1) if nums[i] == left_majority)
            right_count = sum(1 for i in range(left, right + 1) if nums[i] == right_majority)
            
            return left_majority if left_count > right_count else right_majority
        
        return majority_element_rec(0, len(nums) - 1)

File: lab3/lab3-1.java
class Solution {
    public int majorityElement(int[] nums) {
        return majorityElementRec(nums, 0, nums.length - 1);
    }

    private int majorityElementRec(int[] nums, int left, int right) {
        if (left == right) {
            return nums[left];
        }
        
        int mid = (left + right) / 2;
        int leftMajority = majorityElementRec(nums, left, mid);
        int rightMajority = majorityElementRec(nums, mid + 1, right);
        
        if (leftMajority == rightMajority) {
            return leftMajority;
        }
        
        int leftCount = countInRange(nums, leftMajority, left, right);
        int rightCount = countInRange(nums, rightMajority, left, right);
        
        return leftCount > rightCount ? leftMajority : rightMajority;
    }
    
    private int countInRange(int[] nums, int num, int left, int right) {
        int count = 0;
        for (int i = left; i <= right; i++) {
            if (nums[i] == num) {
                count++;
            }
        }
        return count;
    }
}